docs/design/2021-12-09-TiDB-log-based-incremental-backup-draft-1.md

```diff
@@ -0,0 +1,495 @@
+            task.start_ts,
+        );
+    }
+}
+```
+
+Stop the task if end\_ts &lt;= current\_ts.
+
+**Stopping task**
+
+Delete the TaskInfo key. Watcher should be able to do cleanup.
+
+**Configurating, resuming and pausing task**
+
+Update the TaskInfo key. The KV API is equivalent to starting a task.
+Pausing changes the --start-ts to the last key's CommitTS, and set
+"paused" to true. Resuming restores "paused" to false.
+
+**Initialization, leader change and region split/merge**
+
+For every new region, we assume all keys between NextBackupTS:\* to the
+current TS is not yet backed up. So we initiate a full scan on the
+region for CommitTS between these two numbers.
+
+**Flushing**
+
+For every region which this store is still a leader of, update
+NextBackupTS:\* to the ResolvedTS of the region.
+
+**Risks**
+
+-   Still need to sort out all potential crash scenarios
+-   Scanning TxnEntry after changing leaders, no matter when it happened, may reduce the cluster's performance.
+-   Events such as "DeleteRange" and "IngestSST" are not handled.
+-   Should we backup lockcf
+-   How to deal with Rollback (esp across 5min point)
+    -   may lead to dangling default cf entries if the KVDelete is not handled
+    -   in some scenarios there may be a lot of rollbacks?
+    -   The Rollback key will carry start ts. For each key with start ts we need to ensure the delete happened after the put.
+
+## Alternatives & Rationale
+
+**Scanning RaftDB**
+
+Since all modifications must go through Raft consensus, we can capture
+the changes on RaftDB instead of TiKVDB for the incremental changes. We
+can even physically copy the WAL file to minimize computation. Another
+advantage of scanning RaftDB is that, if we want to batch copy events
+between two timestamps (e.g. after changing leader), we don't need to
+scan the entire region key space, but just those entries indexed by TS.
+
+In fact, TiCDC also considered this approach. It was eventually
+abandoned, however, due to the difficulty of actually utilizing the raft
+log. Given the previous failure, we will not attempt to touch the raft
+log unless the TxnEntry scan proved to have a really bad effect on
+performance.
+
+-   Scan the raft log directly. The current raftstore implementation needs to be adjusted, such as raft log gc. In addition, it is also necessary to ensure that the scan is performed during the leader’s lifetime, otherwise the complete data may not be seen.
+
+-   Also need to pay attention to region epoch check, raft log commit != apply successfully
+
+-   To support the interpretation of raftlog, I think I need to implement most of the logic of raftstore now, especially some operations such as split/merge/conf change. Due to the existence of these operations, the original total order raft log in a region has a sloping relationship between different regions. I think it is not easy to solve. This is also the reason why cdc did not use raftlog or wal at the time.
+
+## Trivia
```

Comment:
Parts that only relevant during discussion don't have to be included...

Reply:
Which means we should remove the `trivia` part?

Reply:
yes

Reply:
done